From 59d429bf45aed4e2209043c0c36565ad8e2859a5 Mon Sep 17 00:00:00 2001
From: Umut Cevdet Kocak <umutcevdetkocak@gmail.com>
Date: Mon, 17 Aug 2026 20:37:09 +0300
Subject: [PATCH] screencopy: don't destroy the frame callback installed while
 renegotiating (#425)

---
 src/portals/Screencopy.cpp | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/src/portals/Screencopy.cpp b/src/portals/Screencopy.cpp
index cc232e9..80b860d 100644
--- a/src/portals/Screencopy.cpp
+++ b/src/portals/Screencopy.cpp
@@ -488,10 +488,10 @@ void CScreencopyPortal::SSession::initCallbacks() {
             if ((PSTREAM->pwVideoInfo.format != pwFromDrmFourcc(FMT) && PSTREAM->pwVideoInfo.format != pwStripAlpha(pwFromDrmFourcc(FMT))) ||
                 (PSTREAM->pwVideoInfo.size.width != sharingData.frameInfoDMA.w || PSTREAM->pwVideoInfo.size.height != sharingData.frameInfoDMA.h)) {
                 Debug::log(LOG, "[sc] Incompatible formats, renegotiate stream");
-                sharingData.status = FRAME_RENEG;
-                g_pPortalManager->m_sPortals.screencopy->m_pPipewire->updateStreamParam(PSTREAM);
-                g_pPortalManager->m_sPortals.screencopy->queueNextShareFrame(this);
+                const auto EXPIRED = std::move(sharingData.frameCallback);
                 sharingData.status = FRAME_NONE;
-                sharingData.frameCallback.reset();
+                g_pPortalManager->m_sPortals.screencopy->m_pPipewire->updateStreamParam(PSTREAM);
+                if (!sharingData.frameCallback)
+                    g_pPortalManager->m_sPortals.screencopy->queueNextShareFrame(this);
                 return;
             }
@@ -606,10 +606,10 @@ void CScreencopyPortal::SSession::initCallbacks() {
             if ((PSTREAM->pwVideoInfo.format != pwFromDrmFourcc(FMT) && PSTREAM->pwVideoInfo.format != pwStripAlpha(pwFromDrmFourcc(FMT))) ||
                 (PSTREAM->pwVideoInfo.size.width != sharingData.frameInfoDMA.w || PSTREAM->pwVideoInfo.size.height != sharingData.frameInfoDMA.h)) {
                 Debug::log(LOG, "[sc] Incompatible formats, renegotiate stream");
-                sharingData.status = FRAME_RENEG;
-                g_pPortalManager->m_sPortals.screencopy->m_pPipewire->updateStreamParam(PSTREAM);
-                g_pPortalManager->m_sPortals.screencopy->queueNextShareFrame(this);
+                const auto EXPIRED = std::move(sharingData.windowFrameCallback);
                 sharingData.status = FRAME_NONE;
-                sharingData.windowFrameCallback.reset();
+                g_pPortalManager->m_sPortals.screencopy->m_pPipewire->updateStreamParam(PSTREAM);
+                if (!sharingData.windowFrameCallback)
+                    g_pPortalManager->m_sPortals.screencopy->queueNextShareFrame(this);
                 return;
             }
